R/write.R

```diff
@@ -0,0 +1,364 @@
+
+  # Add custom Sheet # column and label
+  supertbl_toc <- supertbl_toc %>%
+    mutate(`Sheet #` = row_number())
+
+  # Assign label
+  if (labelled) {
+    labelled::var_label(supertbl_toc)$`Sheet #` <- "Sheet #"
+  }
+
+  # Create wb objects
+  wb$add_worksheet(sheet = "Table of Contents")
+  wb$add_data_table(sheet = "Table of Contents",
+                    x = supertbl_toc,
+                    startRow = ifelse(labelled, 2, 1),
+                    tableStyle = table_style)
+  wb$set_col_widths(sheet = "Table of Contents",
+                    cols = seq_along(supertbl_toc),
+                    widths = set_col_widths)
+
+  # Return TOC object as dataframe
+  supertbl_toc
+}
+
+#' @title Add the metadata sheet
+#'
+#' @description
+#' Internal helper function. Adds appropriate elements to wb object. Returns
+#' a dataframe.
+#'
+#' @param supertbl a supertibble generated using `read_redcap()`
+#' @param wb An `openxlsx2` workbook object
+#' @param labelled Whether or not to include labelled outputs, default `FALSE`.
+#' Requires use of `make_labelled()`.
+#' @param table_style Any excel table style name or "none" (see "formatting"
+#' vignette in \link[openxlsx2]{wb_add_data_table}). Default "tableStyleLight8".
+#' @param set_col_widths Width to set columns across the workbook. Default
+#' "auto", otherwise a numeric value. Standard Excel is 8.43.
+#'
+#' @importFrom dplyr select filter
+#' @importFrom tidyr unnest
+#'
+#' @returns A dataframe
+#'
+#' @keywords internal
+
+add_metadata_sheet <- function(supertbl, wb, labelled, table_style, set_col_widths) {
+  supertbl_meta <- supertbl %>%
+    select("redcap_metadata") %>% #nolint: object_usage_linter
+    unnest(cols = "redcap_metadata") %>% # record ID gets duplicated.
+    # Since no other fields are allowed to be duplicated in REDCap,
+    # can use filtering here for removal of duplicated record ID fields
+    filter(!duplicated(.data$field_name))
+
+  wb$add_worksheet(sheet = "REDCap Metadata")
+  wb$add_data_table(sheet = "REDCap Metadata",
+                    x = supertbl_meta,
+                    startRow = ifelse(labelled, 2, 1),
+                    tableStyle = table_style)
+  wb$set_col_widths(sheet = "REDCap Metadata",
+                    cols = seq_along(supertbl_meta),
+                    widths = set_col_widths)
+}
+
+#' @title Check if labelled
+#'
+#' @description
+#' Checks if a supplied supertibble is labelled and throws an error if it is not
+#' but `labelled` is set to `TRUE`
+#'
+#' @param supertbl a supertibble generated using `read_redcap()`
+#' @param labelled Whether or not to include labelled outputs, default `FALSE`.
+#' Requires use of `make_labelled()`.
+#' @param call the calling environment to use in the warning message
+#'
+#' @importFrom cli cli_abort
+#' @importFrom rlang caller_env
+#'
+#' @returns A boolean
+#'
+#' @keywords internal
+
+check_labelled <- function(supertbl, labelled, call = caller_env()) {
```

Comment:
I think this function could be clearer in what it does in each case of what `labelled` is. It seems like the logic is:

```r
# If labelled is FALSE return FALSE
if (!is.null(labelled) && ! labelled) {
    return(FALSE)
}

is_labelled <- # Detect labels

# If labels are there use them
if (is_labelled) {
    return(TRUE)
}

# If user didn't ask for labels that's fine
if (is.null(labelled)) {
    return(FALSE)
} 

# Otherwise error!
```

Reply:
Great, I agree it was pretty confusing as it was written and this is much cleaner.